lib/Notifier.js

```diff
@@ -0,0 +1,371 @@
+
+const logger = new Logger('Notifier');
+
+/**
+ * Termination codes. 
+ */
+const C = {
+  // Termination codes.
+  NOTIFY_RESPONSE_TIMEOUT      : 0, 
+  NOTIFY_TRANSPORT_ERROR       : 1, 
+  NOTIFY_NON_OK_RESPONSE       : 2, 
+  NOTIFY_FAILED_AUTHENTICATION : 3,
+  SEND_FINAL_NOTIFY            : 4, 
+  RECEIVE_UNSUBSCRIBE          : 5, 
+  SUBSCRIPTION_EXPIRED         : 6,
+
+  // Notifer states
+  STATE_PENDING    : 0,
+  STATE_ACTIVE     : 1,
+  STATE_TERMINATED : 2
+};
+
+/**
+ * RFC 6665 Notifier implementation.
+ */
+module.exports = class Notifier extends EventEmitter 
+{
+  /**
+   * Expose C object.
+   */
+  static get C()
+  {
+    return C;
+  }
+
+  get C()
+  {
+    return C;
+  }
+
+  /**
+   * @param {UA} ua - JsSIP User Agent instance.
+   * @param {IncomingRequest} subscribe - Subscribe request.
+   * @param {string} contentType - Content-Type header value.
+   * @param {NotifierOptions} options - Optional parameters.
+   *   @param {Array<string>}  extraHeaders - Additional SIP headers.
+   *   @param {string} allowEvents - Allow-Events header value.
+   *   @param {boolean} pending - Set initial dialog state as "pending". 
+   */
+  constructor(ua, subscribe, contentType, { extraHeaders, allowEvents, pending }) 
+  {
+    logger.debug('new');
+
+    super();
+
+    if (!subscribe)
+    {
+      throw new TypeError('subscribe is undefined');
+    }
+
+    if (!contentType)
+    {
+      throw new TypeError('contentType is undefined');
+    }
+
+    this._ua = ua;
+    this._initial_subscribe = subscribe;
+    this._expires_timestamp = null;
+    this._expires_timer = null;
+
+    // Notifier state: pending, active, terminated. Not used: init, resp_wait.
+    this._state = pending ? C.STATE_PENDING : C.STATE_ACTIVE;
+
+    // Optional. Used to build terminated Subscription-State.
+    this._terminated_reason = null;
+    this._terminated_retry_after = null;
+
+    // Custom session empty object for high level use.
+    this.data = {};
+
+    this._dialog = null;
+    
+    const eventName = subscribe.getHeader('event');
+
+    this._content_type = contentType;
+    this._expires = parseInt(subscribe.getHeader('expires'));
+    this._headers = Utils.cloneArray(extraHeaders);
+    this._headers.push(`Event: ${eventName}`);
+
+    // Use contact from extraHeaders or create it.
+    this._contact = this._headers.find((header) => header.startsWith('Contact'));
+    if (!this._contact)
+    {
+      this._contact = `Contact: ${this._ua._contact.toString()}`;
+
+      this._headers.push(this._contact);
+    }
+
+    if (allowEvents)
+    {
+      this._headers.push(`Allow-Events: ${allowEvents}`);
+    }
+
+    this._target = subscribe.from.uri.user;
+    subscribe.to_tag = Utils.newTag();
+
+    // Create dialog for normal and fetch-subscribe.
+    const dialog = new Dialog(this, subscribe, 'UAS');
+
+    if (dialog.error)
+    {
+      logger.warn(dialog.error);
+
+      throw new Error('SUBSCRIBE missed Contact');
```

Comment:
Why are we throwing this specific error message?

Can't the Contact be checked before creating the dialog?

Reply:
Moved to begin of constructor()

```
    super();

    if (!subscribe)
    {
      throw new TypeError('subscribe is undefined');
    }

    if (!subscribe.hasHeader('contact'))
    {
      throw new TypeError('subscribe - no contact header');
    }

```